contents/handbook/engineering/async-migrations.md

```diff
@@ -0,0 +1,222 @@
+5. Triggers a migration to run if `AUTO_START_SPECIAL_MIGRATIONS` is set and there's an uncompleted migration for this version
+
+#### Running a migration
+
+When a migration is triggered, the following happens:
+
+1. A task is dispatched to Celery to run this migration in the background
+2. The following basic checks are performed to ensure the migration can indeed run:
+    1. We're not over the concurrent migrations limit
+    2. The migration can be run with the current PostHog version
+    3. The migration is not already running
+    4. The service version requirements are met (e.g. X < ClickHouse version < Y)
+    5. The migration's `is_required` check passes
+    6. The migration's `healthcheck` passes
+    7. The migration's dependency (if any) has been completed
+3. We run through each of the operations in order
+4. Every 30 minutes, a Celery task performs a healthcheck, to ensure that:
+   1. The Celery process running the migration didn't crash
+   2. The migration's healthcheck still passes 
+
+> **Note:** Async migrations can also be run synchronously (i.e. not in Celery) using the async migrations CLI (WIP) or the Django shell.
+
+#### Stopping a migration
+
+A migration can be stopped by issuing a command via Celery's app control to terminate the process running the task. 
+
+#### Rollbacks
+
+If a migration is stopped for any reason (manual trigger or error), we will attempt to roll back the migration following the operations specified in reverse order from the last started operation. 
+
+If a roll back succeeds, the migration status will be updated to reflect this.
+
+#### Errors
+
+If a migration errors, the error message is added to the migration's database record and we automatically trigger a rollback.
+
+### Scope and limitations
+
+The initial implementation of async migrations targets only **data migrations**, and assumes that the migration is used as a mechanism to help users move into a new default state. 
+
+For example, when we [moved our ClickHouse `person_distinct_id` table to a `CollapsingMergeTree`](https://github.com/PostHog/posthog/pull/5563), we updated the SQL for creating the table, and wrote a migration to help users on the old schema migrate to the new schema. 
+
+However, users that did a fresh deploy of PostHog _after_ this change already had the table with the new schema created by default.
+
+This is the only type of operation that async migrations _currently_ support, to prevent a complex web of dependencies between migration types.
+
+As such, those writing an async migration should write a sensible `is_required` function that determines if the migration should run or not. 
+
+Thus, when a user deploys a new PostHog instance, we will first run **all** EE migrations in order, and then **all** of the async migrations in order. At this step, async migrations should be skipped if the codebase already contains updated default schemas. 
```

Comment:
please explain how the codebase can already contain the updated default schemas, how does that work? Is the way that we'd write the same the special migration as a normal migration after the version upper bound is passed (via if table doesn't exist etc)?

Reply:
Take a look at the PR linked above. We updated the base SQL for the table and wrote a migration to get users on it. It's the normal way we do CH migrations actually.